common/log.c

```diff
@@ -634,6 +694,149 @@ log_end(void)
     return ret;
 }
 
+/*****************************************************************************/
+/* produce a hex dump */
+enum logReturns
+log_hexdump_with_location(const char *function_name, 
+                          const char *file_name, 
+                          const int line_number, 
+                          const enum logLevels log_level, 
+                          const char *message, 
+                          const char *src, 
+                          int len)
+{
+    unsigned char *line;
+    int i;
+    int dump_number_lines;
+    int dump_line_length;
+    int dump_header_length;
+    int dump_length;
+    int dump_offset;
+    int thisline;
+    int offset;
+    char *dump_buffer;
+    enum logReturns rv;
+    bool_t override_destination_level = 0;
+
+    override_destination_level = internal_log_location_overrides_level(function_name, 
+                                                                       file_name, 
+                                                                       log_level);
+    if(!internal_log_is_enabled_for_level(log_level, override_destination_level))
+    {
+        return LOG_STARTUP_OK;
+    }
+
+    dump_line_length = 4 + 3             /* = 4 offset + 3 space */
+                       + ((2 + 1) * 16)  /* + (2 hex + 1 space) per byte for 16 bytes */
+                       + 2 + 16 + 1;     /* + 2 space + 16 bytes as chars + 1 newline */
+    dump_header_length = 12 + 1; /* "%s Hex Dump:" header + newline */
+#ifdef _WIN32
+    dump_line_length += 1;               /* + 1 WIN32 newline */
+    dump_header_length += 1;             /* + 1 WIN32 newline */
+#endif
+    dump_number_lines = (len / 16) + 1; /* 16 bytes per line then round up */
+    dump_length = dump_number_lines * dump_line_length /* hex dump space */
+                    + dump_header_length
+                    + 1;    /* terminating NULL */
+    dump_buffer = (char *)g_malloc(dump_length, 1);
+    line = (unsigned char *)src;
+    offset = 0;
+
+    /* Start the dump with a new line so that the first line of the dump is 
+       aligned to the first column instead of to after the log message 
+       preamble (eg. time, log level, ...)
+    */
+    g_sprintf(dump_buffer, "%%s Hex Dump:");
```

Comment:
Extra comment which I managed to lose in the review:-

Suggest making `"%s Hex Dump:\n"` a macro like this:-

```
#ifdef _WIN32
#define HEX_DUMP_HEADER "%s Hex Dump:\r\n"
#else
#ifdef _MACOS
#define HEX_DUMP_HEADER "%s Hex Dump:\r"
#else
#define HEX_DUMP_HEADER "%s Hex Dump:\n"
#endif
#endif
#define HEX_DUMP_HEADER_SIZE (sizeof(HEX_DUMP_HEADER) - 1)
```

I'm sure you get the idea. You can them remove the magic number 12.

Reply:
agreed, fixed